### DISCUSSION TITLE
lit-element with scss

### DISCUSSION THREAD

**Discussion by klemenoslaj** [parcel-bundler/parcel#4640] [General] [answered]
I think the title explains what I am about to ask.

`lit-element` works great out of the box. I can pair it with JS or TS and all works well.
I wonder what would I require to compile scss instead of using pure css with `lit-element`?

**Comment by mischnic** [comment, marked as answer]
I'm not familiar with lit-element, but I think you want to compile SCSS and get the result as a string inside JS? (If not, please give an example of what you're trying to achieve).

With Parcel 2, you can do this out of the box (the example at the very bottom uses native html custom elements which should be rather similar): https://parcel2-docs.now.sh/getting-started/plugin-config/#predefined-(offical)-named-pipelines

**Reply by klemenoslaj** [reply]
Thanks for the reply.

I will try parcel2 today, thanks. Last time I tried I had trouble with decorators even though I configured them in `tsconfig.json`.

**Reply by mischnic** [reply]
> I had trouble with decorators even though I configured them in tsconfig.json.

Maybe related to this https://github.com/parcel-bundler/parcel/issues/4544#issuecomment-622056701?

**Reply by klemenoslaj** [reply]
Ah so by default parcel2 uses babel to transpile typescript?
This setting tells parcel to use tsc transformer instead, right?

**Reply by mischnic** [reply]
Correct (and Babel also completely ignores tsconfig.json)

**Reply by klemenoslaj** [reply]
Yes I know. To my understanding it only enables TS syntax.
Why parcel uses babel instead of tsc tho?

---

Sorry this is now all unrelated to the original question. Feel free to just ignore it from here.
I'll mark as answered once I get to test it later tonight. Thanks again :+1:

**Reply by mischnic** [reply]
> Why parcel uses babel instead of tsc tho?

It's generally faster, here's the discussion around that: https://github.com/parcel-bundler/parcel/issues/4022

**Reply by klemenoslaj** [reply]
Yes, it's faster, because it doesn't transpile typescript at all. To my understanding it just strips it out. So type checker is not even triggered, as far as I know.

If that's true, is that a good default configuration for parcel I wonder? 🤔

**Reply by klemenoslaj** [reply]
https://github.com/parcel-bundler/parcel/issues/4022#issuecomment-576330089

Couldn't have said that better.

**Reply by klemenoslaj** [reply]
Unfortunately I cannot make parcel2 work. Current config:

```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.ts": [
      "@parcel/transformer-typescript-tsc"
    ]
  }
}
```

With `@next` I get this funny error:

```
Error: Unexpected token 'export'
/home/user/projects/base/node_modules/@parcel/ts-utils/src/index.js:2
export * from './FSHost';
^^^^^^
````

This is apparently fixed in nightly, but nightly fails too with different error:

```
Error: Cannot find module '@parcel/config-default' from '/home/user/projects/base'
```

I tried installing config-default on my own, but that didn't help.

---

I am kind of locked out, but I will accept your answer since it's correct. I can compile scss separately and import string.

**Reply by klemenoslaj** [reply]
Ok I managed to get it running, but I had to configure `.parcelrc` for `*.ts`, `*.html` and `*.scss` to make it work as nightly parcel couldn't find `@parcel/config-default`, and `next` parcel had a problem importing `./fHost`.

---

Now this works, however it looks quite bad.

With `css`:

```ts
@customElement('my-element')
export default MyElement extends LitElemnet {
  static styles = css`
    :host {
      background: red;
    }
  `;
}
```

With `scss`:

```ts
import style from 'bundle-text:./my-element.scss';

@customElement('my-element')
export default MyElement extends LitElemnet {
  static styles = css({ raw: [style], ...[style] });
}
```

---

Is it possible for me to write my own transformer that uses scss transformer and outputs something like this:

```ts
import { css } from 'lit-element';

export const style = css`<CSS CONTENT FROM SCSS>`;
```

**Reply by mischnic** [reply]
> I tried installing config-default on my own, but that didn't help.

That's very strange. Do you have time to create a repo where this happens?

> Is it possible for me to write my own transformer that uses scss transformer and outputs something like this:

Yes. It would be conceptionally be similar to how inline CSS style stags are transformed:
`<style> some code </style>` is turned into `<style>some_random_hash</style>` and then later on the packager replaces this with the bundle contents.

So your transformer would return additional assets (one per `css` template string) where `isInline: true` and with a `uniqueKey` (e.g. one counter per file) and the contents (`<CSS CONTENT FROM SCSS>`). Then you need to replace `<CSS CONTENT FROM SCSS>` in the JS Asset with the return value of `asset.addDependency(theUniqueKeyFromBefore)` (some_random_hash from before). The default JSPackager should then replace the hash without any changes. (We will have docs for writing plugins very soon, I'll make sure to add a section about how inline bundles work.)

**Reply by klemenoslaj** [reply]
> That's very strange. Do you have time to create a repo where this happens?

Sure, I'll make one tomorrow. Will try to reproduce it on a clean setup, just to eliminate the additional possibilities of failure. I wonder if `.parcel-cache` could be corrupt tho, as I do not recall deleting it, but I did delete it later.

> Yes. It would be conceptionally be similar to how inline CSS style stags are transformed:

Awesome! I will try to come up with something then. Looking forward to dig into parcel - it's amazing, good job guys :+1:

**Reply by klemenoslaj** [reply]
Ok I just tried to reproduce this situation in a clean repository. While I succeeded it seemed to have been fixed after I removed `node_modules` and `.parcel-cache` and reinstalled.

Worth mentioning I am using `yarn`. Could it be that your auto-installation process messes up something?

---

Next thing I noticed, `@parcel/transformer-react-refresh-babel` got installed even though there is no notion of react in my setup? Expected?

---

Side question: did you consider batching all the installations into one?

**Reply by mischnic** [reply]
> Next thing I noticed, @parcel/transformer-react-refresh-babel got installed even though there is no notion of react in my setup? Expected?

Yes, it's part of the default config so that React HMR works without a custom config.
It's however a noop if you don't have react in your package.json dependencies:
https://github.com/parcel-bundler/parcel/blob/d33fcd9cf826188288e7ffb5dfecbfa221592458/packages/transformers/react-refresh-babel/src/ReactRefreshBabelTransformer.js#L16

> Side question: did you consider batching all the installations into one?

That does make sense, but it's also possible that there is a chain of autoinstalls (so a second autoinstall is only triggered after the first one finished and the code that used the package ran - e.g. a pug file that includes a scss stylesheet).

**Reply by klemenoslaj** [reply]
> That does make sense, but it's also possible that there is a chain of autoinstalls (so a second autoinstall is only triggered after the first one finished and the code that used the package ran - e.g. a pug file that includes a scss stylesheet).

Ok, that makes it more complex unless there is a way of predicting the chain.

---

I pushed my repo anyway: https://github.com/klemenoslaj/parcel-nightly-test

Please note it is without `yarn.lock` intentionally so the installation process goes from beginning if you try it out. As well leaves room to test `npm` too without compatibility issues of some sorts.

Another thing...after manual reinstallation the config default error is gone, however the error of `fHost` appears. That's due to `transformer-typescript-tsc` version `2.0.0-alpha.3` being installed automatically instead of `nightly-277`. installing `nightly-277` manually resolves this.
